### ISSUE TITLE
New rule suggestion: Kotlin files containing only top-level declarations should match a naming convention

### ISSUE THREAD

**Issue by drawers** [detekt/detekt#7529]
## Expected Behavior of the rule

Let there be two rules:
1. A rule for addressing a general case where there is a Kotlin file with only top-level declarations. The rule checks if such files correspond to a configurable regex e.g., `"^[A-Z][a-z0-9]+(?:[A-Z][a-z0-9]+)*s$"` (Pascal Case ending in 's') 

Compliant ✅
```kotlin
// Extensions.kt
fun MyClass.process() = // …
fun MyResult.print() = // …
```

Non-compliant ❌
```kotlin
// extensions.kt
fun MyClass.process() = // …
fun MyResult.print() = // …
```

Non-compliant ❌
```kotlin
// Util.kt
fun MyClass.process() = // …
fun MyResult.print() = // …
```

2. A rule for addressing a specific case where there is a file where the only public/internal top-level declarations have a specific type as the receiver. 

In this case, the user can use a YAML string config to choose a convention from a number of options (see Twitter poll below) e.g., `Strings.kt`, `Offsets.kt` for extensions on `String.kt` and `Offset.kt`:

Compliant ✅
```
// Strings.kt
fun String.html(context: Context) = // …
fun String.print() = // …
```

Compliant ✅
```
// Strings.kt
internal fun String.html(context: Context) = // …
internal fun String.print() = // …
```

Non-compliant ❌
```kotlin
// String.kt
internal fun String.html(context: Context) = // …
internal fun String.print() = // …
```

Non-compliant ❌
```kotlin
// stringUtils.kt
fun String.html(context: Context) = // …
fun String.print() = // …
```

## Context
Kotlin code bases often accrue Kotlin files containing only extension functions and extension properties for a given type. For example, the [StringExtensions.kt](https://github.com/duckduckgo/Android/blob/c988936e0f662de44648e5847afb4143d0cfa72f/common/common-utils/src/main/java/com/duckduckgo/common/utils/extensions/StringExtensions.kt) file from the DuckDuckGo codebase that contains extension functions on `String` like:

```kotlin
fun String.html(context: Context): Spanned {
    return Html.fromHtml(this, Html.FROM_HTML_MODE_COMPACT, { htmlDrawable(context, it.toInt()) }, null)
}

private fun htmlDrawable(
    context: Context,
    resource: Int,
): Drawable? {
    return ContextCompat.getDrawable(context, resource)?.also {
        it.setBounds(0, 0, it.intrinsicWidth, it.intrinsicHeight)
    }
}
```

There are a number of different conventions for naming these files, as explained in [this Twitter poll](https://x.com/cyrilmottier/status/1806368965905551464):

![Image of a poll from Twitter](https://github.com/user-attachments/assets/4e164d1c-9303-440b-831d-adbb1b7ac1be)

The [Kotlin Coding conventions](https://kotlinlang.org/docs/coding-conventions.html#source-file-names) has this to say on the matter:

>If a Kotlin file contains a single class or interface (potentially with related top-level declarations), its name should be the same as the name of the class, with the .kt extension appended. It applies to all types of classes and interfaces. If a file contains multiple classes, or only top-level declarations, choose a name describing what the file contains, and name the file accordingly. Use [an upper camel case](https://en.wikipedia.org/wiki/Camel_case) with an uppercase first letter (also known as Pascal case), for example, ProcessDeclarations.kt.

>The name of the file should describe what the code in the file does. Therefore, you should avoid using meaningless words such as Util in file names.

This is [interpreted by Ktlint](https://github.com/pinterest/ktlint/issues/1521) to mean that top-level only files should be named in Pascal case.

But the [Android Kotlin style guide](https://developer.android.com/kotlin/style-guide) recommends lower-case `extensions.kt` for this sort of thing:

```kotlin
// extensions.kt
fun MyClass.process() = // …
fun MyResult.print() = // …
```

## Prior art

Ktlint has [the following rule](https://github.com/pinterest/ktlint/blob/master/ktlint-ruleset-standard/src/main/kotlin/com/pinterest/ktlint/ruleset/standard/rules/FilenameRule.kt) but the philosophy is slightly different. It's not configurable and it doesn't specify a `*Utils.kt` or `*Extension.kt` convention for a given type.

Detekt has [MatchingDeclarationName](https://github.com/detekt/detekt/blob/main/detekt-rules-naming/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/naming/MatchingDeclarationName.kt) but the intention there is to make sure that the file name matches a top-level class or object declaration.

## Other considerations

In some sense, these proposed new rules rush in where angels fear to tread since the Kotlin coding conventions don't specify anything other than PascalCase for Kotlin files only containing top-level declarations. There is no set convention `*Utils.kt` or `*Extensions.kt` either.

There's also the case where a developer starts to write a file that is intended to contain extensions on multiple types but only adds one declaration. Proposed rule 2 forces them to name the file after the receiver of the extension function even when they're not ready (although this could be suppressed).

If the rules are implemented, they should also handle multi-platform conventions (multiplatform target suffixes like "jvm")

**Comment by drawers** [comment]
Related: https://github.com/detekt/detekt/issues/1500

**Comment by cortinico** [comment]
This is an interesting rule as I stumbled upon this type of conversation multiple times.
We should definitely align on the convention to enforce though.

**Comment by atulgpt** [comment]
Regarding `But the [Android Kotlin style guide](https://developer.android.com/kotlin/style-guide) recommends lower-case extensions.kt for this sort of thing` they say it is acceptable though their first recommendation is to use `PascalCase`

> Otherwise, if a source file contains multiple top-level declarations, choose a name that describes the contents of the file, apply PascalCase (camelCase is acceptable if the filename is plural),

**Comment by drawers** [comment]
> Android Kotlin style guide

Thanks, @atulgpt  that's a really good point. I think it would be desirable for teams to be able to encode the Android Kotlin style guide using Detekt. On the other hand, that convention "camelCase is acceptable if the file name is plural" is hard to enforce using static analysis since we have no way of distinguishing a plural from a mere singular that happens to end in 's'. 

One way forward is to abandon the first proposed rule (the generic one that focuses on any file that contains only top-level declarations) and concentrate on the second. This has the advantage of focusing on the second rule (what to call "FooExtensions.kt" and "FooUtils.kt"). In my experience, this second one is more of a site for bike-shedding. 

What do you think?
